content/blog/blog-2020-07-building-your-first-operator-2.md

```diff
@@ -0,0 +1,633 @@
+```
+
+Before we break this code down, let's clarify a couple of assumptions that we can make:
+
+1. Our nodes will be in a StatefulSet, so we can predict the naming of the pods will be `.Name-.OperatorName-number`, where `.Name` and `.OperatorName` are variables created by KUDO and number will be assigned by Kubernetes based on the order the replicas in the StatefulSet were instantiated.
+2. We will be creating a headless service to access them, which we will name as `.Name-hs` to ensure uniqueness, and we will document that later in this blog. That will give us predictable DNS entries for each node in our StatefulSet in the format `.Name-.OperatorName-number.Name-hs`
+
+Working from those assumptions we can start to put together the connection string. 
+
+Firstly we add the service address for our bootstrap node, which we created in the last blog, followed by a comma:
+
+```
+gcomm://{{ .Name }}-bootstrap-svc,
+```
+
+We also know there will always be a node ending in 0, since we must have at least 1 node, so we can add the DNS entry for that node's service:
+
+```
+gcomm://{{ .Name }}-bootstrap-svc,{{ $.Name }}-{{ $.OperatorName }}-0.{{ $.Name }}-hs
+```
+
+For the next part, we need to define a value for the number of nodes we are going to deploy, in our `params.yaml` so we can refer to that here :
+
+```yaml
+# params.yaml
+  - name: NODE_COUNT
+    description: "Number of nodes to create in the cluster"
+    default: "3"
+```
+
+Now we can use that variable in the next step of the Sprig templating :
+
+```
+{{ range $i, $v := untilStep 1 (int .Params.NODE_COUNT) 1 }},{{ $.Name }}-{{ $.OperatorName }}-{{ $v }}.{{ $.Name }}-hs{{end}}
+```
+
+Here we are using the range function, part of standard Go templating, which gives us values and indexes in `$v` and `$i`, to which we are passing a range generated using the UntilStep function from Sprig. UntilStep called in this way will give us a list of integers, starting at 1, up to `NODE_COUNT`, in steps of 1. So if `NODE_COUNT` is 3, then the range generated by UntilStep here would be 1,2. 
+
+Note that we have to cast `.Params.NODE_COUNT` to an integer before we use it, since it’s just a string in our `params.yaml` file. We also need to precede the other string variables we are accessing inside our function with `$`, to refer to the top-level context again, since when the range function is invoked, variables are re-scoped to within the function.
+
+Then this code will print out each entry for the nodes in our cluster, preceded by a comma. Finally we close the templating code with `{{end}}`
+
+Working out complex templating functions can be tricky, you might find it useful to have some stub Go code you can run when testing the templating. I have some [example code](https://github.com/mattj-io/template_test) in Github, which just prints out the template to stdout. 
+
```

Comment:
```suggestion
::: tip
You can always run `kubectl kudo package verify .` to ensure that your templating is syntactically correct. This won't show you any logical mistakes, but it will catch a couple of problems.
:::
```